lighthouse-core/gather/driver.js

```diff
@@ -630,6 +630,60 @@ class Driver {
     return new LHElement(targetNode, this);
   }
 
+  /**
+   * @param {string} handle
+   * @param {{timeout: number}=} options,
+   * @return {Promise<string>}
+   */
+  async readIOStream(handle, options = {timeout: 5000}) {
+    const startTime = Date.now();
+
+    let ioResponse;
+    let data = '';
+    while (!ioResponse || !ioResponse.eof) {
+      if (Date.now() - startTime > options.timeout) {
+        throw new Error('Waiting for the end of the IO stream exceeded the allotted time.');
+      }
+      ioResponse = await this.sendCommand('IO.read', {handle});
+      data = data.concat(ioResponse.data);
+    }
+
+    if (ioResponse.base64Encoded) {
+      data = Buffer.from(data, 'base64').toString('utf-8');
+    }
+    return data;
+  }
+
+  /**
+   * @param {string} url
+   * @return {Promise<string>}
+   */
+  async fetchResourceOverProtocol(url) {
+    const milestone = await this.getBrowserVersion().then(v => v.milestone);
+    if (milestone < 87) {
+      throw new LHError(LHError.errors.UNSUPPORTED_OLD_CHROME, {
+        featureName: 'Network.loadNetworkResource',
+      });
+    }
+
+    const frameTreeResponse = await this.sendCommand('Page.getFrameTree');
+    const networkResponse = await this.sendCommand('Network.loadNetworkResource', {
+      frameId: frameTreeResponse.frameTree.frame.id,
+      url,
+      options: {
+        disableCache: true,
+        includeCredentials: true,
+      },
+    });
+
+    if (!networkResponse.resource.success || !networkResponse.resource.stream) {
+      const statusCode = networkResponse.resource.httpStatusCode || '';
+      throw new Error(`Loading network resource failed (${statusCode})`);
```

Comment:
Should we handle this error differently in `SourceMaps`? In this case, the old fetcher will likely fail as well so maybe we shouldn't spend time falling back.

Reply:
This could be addressed by using the chrome version check to determine _which_ implementation to use, instead of switching to the other based on any failure.

I think keeping just one exported function (the existing `driver.fetcher.fetchResource`), and having that function do a chrome version check and branch on impl. would make the calling code more straightforward.